core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,137 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import org.apache.iceberg.exceptions.ValidationException;
+
+public class SnapshotManager extends MergingSnapshotProducer<ManageSnapshots> implements ManageSnapshots {
+
+  private final TableOperations ops;
+  private SnapshotManagerOperation operation;
+  private TableMetadata base;
+  private Long targetSnapshotId = null;
+
+  enum SnapshotManagerOperation {
+    CHERRYPICK,
+    ROLLBACK
+  }
+
+  SnapshotManager(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.current();
+  }
+
+  @Override
+  protected ManageSnapshots self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollback(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.ROLLBACK;
+    this.targetSnapshotId = snapshotId;
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackAtTime(long timestampMillis) {
+    long snapshotId = 0;
+    long snapshotTimestamp = 0;
+    operation = SnapshotManagerOperation.ROLLBACK;
+    // find the latest snapshot by timestamp older than timestampMillis
+    for (Snapshot snapshot : base.snapshots()) {
+      if (snapshot.timestampMillis() < timestampMillis &&
+          snapshot.timestampMillis() > snapshotTimestamp) {
+        snapshotId = snapshot.snapshotId();
+        snapshotTimestamp = snapshot.timestampMillis();
+      }
+    }
+
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back, no valid snapshot older than: %s", timestampMillis);
+
+    this.targetSnapshotId = snapshotId;
+
+    return this;
+  }
+
+  @Override
+  public Snapshot apply() {
+    // common checks
+    ValidationException.check(targetSnapshotId != null,
+        "Cannot run operations on an unknown version: call rollback, rollbackAtTime or cherrypick");
+    ValidationException.check(operation != null,
+        "Need to define operation on snapshot: call rollback, rollbackAtTime or cherrypick");
+
+    if (operation == SnapshotManagerOperation.CHERRYPICK) {
```

Comment:
Minor: For enums, I think we prefer using `switch` statements so that we don't need to run each branch condition separately.

Reply:
+1